File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    #Brute force Bozo Solution:
    # Interesting: Doing an in-order traversal always gives items in a sorted order from a BST!
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        entries = []
        def dfs(root : TreeNode):
            if root is None:
                return None
            dfs(root.left)
            entries.append(root.val)
            dfs(root.right)
        
        dfs(root)
        return entries[k-1]

File: Data Structures & Algorithms/invert-a-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
//Iterative DFS Approach
class Solution {
    public TreeNode invertTree(TreeNode root) {
        //Handle base case
        //Initialise Stack
        //Push L and Right nodes appropritely.
        if(root == null){
            return null;
        }
        Stack<TreeNode> stack_ = new Stack<>();
        stack_.push(root);
        while(!(stack_.isEmpty())){
            TreeNode topMost = stack_.pop();
            TreeNode temp = topMost.right;
            topMost.right = topMost.left;
            topMost.left = temp;
            if(topMost.left != null){
                stack_.push(topMost.left);
            }
            if(topMost.right != null){
                stack_.push(topMost.right);
            }
        }
        return root;
    }
}
